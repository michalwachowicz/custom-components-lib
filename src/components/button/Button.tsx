import {
  MouseEventHandler,
  useEffect,
  useLayoutEffect,
  useState,
  useRef,
} from "react";
import * as styles from "./Button.module.scss";

export type ButtonVariant = "text" | "contained" | "outlined";
export type ButtonSize = "small" | "medium" | "large";

interface ButtonProps {
  variant?: ButtonVariant;
  size: ButtonSize;
  disabled?: boolean;
  children: React.ReactNode;
  ref?: React.RefObject<HTMLButtonElement | null>;
  onClick?: MouseEventHandler<HTMLButtonElement>;
}

interface Entity {
  id: number;
  x: number;
  y: number;
}

function ButtonEntity({ size, entity }: { size: number; entity: Entity }) {
  const { x, y } = entity;
  const sizePx = `${size}px`;

  return (
    <span
      data-testid='btn-entity'
      className={styles["btn-entity"]}
      style={{
        height: sizePx,
        width: sizePx,
        top: `${y}px`,
        left: `${x}px`,
      }}
    />
  );
}

export default function Button({
  variant = "contained",
  disabled = false,
  size,
  children,
  ref,
  onClick = () => {},
}: ButtonProps) {
  const btnRef = useRef<HTMLButtonElement>(null);
  const finalRef = ref || btnRef;

  const [entity, setEntity] = useState<Entity | null>(null);
  const [height, setHeight] = useState(0);

  const spawnEntity = (x: number, y: number) => {
    setEntity({ id: Math.random(), x, y });
  };

  const clickHandler: MouseEventHandler<HTMLButtonElement> = (e) => {
    if (disabled) return;
    if (onClick) onClick(e);

    spawnEntity(e.clientX, e.clientY);
  };

  useLayoutEffect(() => {
    setHeight(finalRef.current?.offsetHeight || 0);
  }, [finalRef]);

  useEffect(() => {
    if (entity !== null) {
      const timer = setTimeout(() => setEntity(null), 500);
      return () => clearTimeout(timer);
    }
  }, [entity]);

  return (
    <button
      ref={finalRef}
      className={`${styles.btn} ${styles[`btn-${variant}`]} ${
        styles[`btn-${size}`]
      }`}
      type='button'
      disabled={disabled}
      onClick={clickHandler}
    >
      {children}
      <span className={styles["btn-backdrop"]}>
        {entity !== null && (
          <ButtonEntity key={entity.id} size={height} entity={entity} />
        )}
      </span>
    </button>
  );
}
