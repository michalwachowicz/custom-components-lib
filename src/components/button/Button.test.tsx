import { act } from "react";
import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Button, { ButtonSize, ButtonVariant } from "./Button";

describe("<Button />", () => {
  const sizes = ["small", "medium", "large"];
  const variants = ["text", "contained", "outlined"];
  const disabledStates = [true, false];

  it("renders children properly", () => {
    render(<Button size='medium'>Test</Button>);
    expect(screen.getByRole("button")).toHaveTextContent("Test");
  });

  it("sets disabled status properly", () => {
    const { rerender } = render(<Button size='medium'>Test</Button>);
    expect(screen.getByRole("button")).not.toBeDisabled();

    rerender(
      <Button size='medium' disabled>
        Test
      </Button>
    );
    expect(screen.getByRole("button")).toBeDisabled();
  });

  describe("applies the correct size class", () => {
    test.each(sizes)('applies "btn-%s" class when size="%s"', (size) => {
      render(<Button size={size as ButtonSize}>Test</Button>);
      const btn = screen.getByRole("button");
      expect(btn).toHaveClass(`btn-${size}`);
    });
  });

  describe("applies the correct variant class", () => {
    test.each(variants)(
      'applies "btn-%s" class when variant="%s"',
      (variant) => {
        render(
          <Button variant={variant as ButtonVariant} size='medium'>
            Test
          </Button>
        );
        const btn = screen.getByRole("button");
        expect(btn).toHaveClass(`btn-${variant}`);
      }
    );
  });

  describe("renders correctly with different props and matches snapshot", () => {
    const testCases = sizes.flatMap((size) =>
      variants.flatMap((variant) =>
        disabledStates.map((disabled) => [size, variant, disabled])
      )
    );

    test.each(testCases)(
      "renders correctly with size=%s, variant=%s, disabled=%s",
      (size, variant, disabled) => {
        const { asFragment } = render(
          <Button
            size={size as ButtonSize}
            variant={variant as ButtonVariant}
            disabled={!!disabled}
          >
            Test
          </Button>
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });

  it("runs click handler and spawns entity", async () => {
    jest.useFakeTimers();

    const onClick = jest.fn();
    const user = userEvent.setup({ delay: null });

    render(
      <Button size='medium' onClick={onClick}>
        Test
      </Button>
    );

    expect(onClick).toHaveBeenCalledTimes(0);
    expect(screen.queryByTestId("btn-entity")).not.toBeInTheDocument();

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(screen.getByTestId("btn-entity")).toBeInTheDocument();

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.queryByTestId("btn-entity")).not.toBeInTheDocument();
    jest.useRealTimers();
  });

  it("does not run when disabled", async () => {
    const onClick = jest.fn();
    const user = userEvent.setup({ delay: null });

    render(
      <Button size='medium' onClick={onClick} disabled>
        Test
      </Button>
    );

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    expect(onClick).toHaveBeenCalledTimes(0);
    expect(screen.queryByTestId("btn-entity")).not.toBeInTheDocument();
  });
});
