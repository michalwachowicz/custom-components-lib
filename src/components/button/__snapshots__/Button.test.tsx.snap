// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=contained, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-large"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=contained, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-large"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=outlined, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-large"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=outlined, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-large"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=text, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-large"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=large, variant=text, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-large"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=contained, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-medium"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=contained, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-medium"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=outlined, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-medium"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=outlined, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-medium"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=text, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-medium"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=medium, variant=text, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-medium"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=contained, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-small"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=contained, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-contained btn-small"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=outlined, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-small"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=outlined, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outlined btn-small"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=text, disabled=false 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-small"
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with different props and matches snapshot renders correctly with size=small, variant=text, disabled=true 1`] = `
<DocumentFragment>
  <button
    class="btn btn-text btn-small"
    disabled=""
    type="button"
  >
    Test
    <span
      class="btn-backdrop"
    />
  </button>
</DocumentFragment>
`;
